Proposal for recursive constraints on associated types

austinzheng opened the pull request:
This isn't 100% ready to go, but I figure reviewing a pull request might be more convenient than comments through email and directly on commits. I've incorporated a round of feedback from @dabrahams, @DougGregor, and @erica.

Once this is finished I'll squash the commits and fix the commit message.

TODO: fill out the missing stdlib changes.

austinzheng commented on proposals/XXXX-recursive-protocol-constraints.md:
````diff
@@ -0,0 +1,265 @@
+semantically incorrect, and would be rejected by a Swift compiler implementing this change.
+
+```swift
+struct BadSequence : Sequence, IteratorProtocol {
````
I'll improve the example. I agree it's pretty stupid; I couldn't think of anything better.

airspeedswift commented:
Hi Austin – do you think you'll be able to cover the requested changes? Would be good to get this into review if we can resolve these.

austinzheng commented:
Sorry @airspeedswift, I've been horribly busy these past few days but will definitely carve out some time tomorrow or Saturday to address your points. Apologies for the delays (again).

natecook1000 commented on proposals/XXXX-recursive-protocol-constraints.md:
```diff
@@ -0,0 +1,265 @@
+
+### `Sequence`
+
+* Change associated type: `associatedtype SubSequence : Sequence where Iterator.Element == SubSequence.Iterator.Element, SubSequence.SubSequence == SubSequence`
```
I don't think we need to add the constraint `SubSequence.SubSequence == SubSequence`—this is generally true in practice, but I think we're using it currently as a constraint because it makes the other constraints endlessly recursive. Once the constraints are in place on the protocols we should only need to require that the subsequence is a sequence with the same `Element` type.

dabrahams replied:
This is definitely a constraint I want to keep.
We're using it because it makes

```swift
var x = a[a.startIndex..<a.endIndex]
while !x.isEmpty {
   ...
   x = x.dropFirst()
}
```

possible, not because of any infinite recursion.

natecook1000 replied:
Sounds good, withdrawn!

dabrahams commented on proposals/XXXX-recursive-protocol-constraints.md:
```diff
@@ -0,0 +1,265 @@
+
+* Remove conformance to `_Indexable`
+* Change associated type: `associatedtype SubSequence : Collection where SubSequence.Index == Index, SubSequence.Indices == Indices`
+* Change associated type: `associatedtype Indices : Collection where Indices.Iterator.Element == Index, Indices.Index == Index, Indices.SubSequence == Indices`
```
That constraint seems unnecessary to me, too.  Does it come from some code somewhere?

austinzheng replied:
I believe I copied it out of a `FIXME` comment. I'll check for sure when I get home tonight.

airspeedswift replied:
from here, presumably: https://github.com/apple/swift/blob/master/stdlib/public/core/Collection.swift#L648

airspeedswift replied:
but I agree it doesn't seem necessary. `Indices.Element == Index` and `SubSequence.Index == Index` ought to be enough

dabrahams replied:
It's interesting... I can't think of a model of `Collection` whose `Indices` couldn't be its own subsequence.  It's a small-ish programming convenience, but there's something to be said for locking it down.

gribozavr replied:
I think I added this constraint because I thought it was necessary for some implementation detail of the generic `Slice` type.

natecook1000 commented:
Does this protocol allow us to require that an associated type is equal to `Self`? For example, the `FloatingPoint` protocol should perhaps require that `Stride == Self`.

austinzheng commented:
Thank you, all, for your patience. I've addressed the comments.

I believe the primary outstanding work now is adding more standard library APIs that need to be changed in response to the new features. Is that correct?

austinzheng commented:
> Does this protocol allow us to require that an associated type is equal to `Self`?

I'm not sure I understand the point of doing this, @natecook1000. Naively, I would think you'd just use `Self` anywhere you would otherwise use an associated type constrained to be equal to `Self`. Am I missing something?

austinzheng commented:
Just wanted to let @airspeedswift know that this has been updated. There is no rush on my part to get it in, so feel free to review it whenever you have time.

rjmccall commented:
Taking.

austinzheng commented:
If there are any further changes you'd like me to make, I'd be happy to make them and submit a follow-up PR (in a timely manner).

rjmccall commented:
Unless it's editorial, or so serious that we should halt and re-start the review, it should probably just get discussed on the list.
